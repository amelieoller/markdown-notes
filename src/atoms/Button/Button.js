import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

const withOnText = (string) => `on${string.charAt(0).toUpperCase() + string.slice(1)}`;

const propsCSS = {
  outline: css`
    background-color: transparent;
    color: ${({ theme }) => theme.colors.onSurface};
    border: 2px solid
      ${({ theme, buttonType }) =>
        buttonType ? theme.colors[buttonType] : theme.colors.secondary};

    &:hover {
      background-color: ${({ theme, buttonType }) =>
        buttonType ? theme.colors[buttonType] : theme.colors.secondary};
      color: ${({ theme, buttonType }) =>
        buttonType ? theme.colors[withOnText(buttonType)] : theme.colors.onSecondary};
    }
  `,

  small: css`
    font-size: 1.1rem;
    padding: 4px 9px;

    svg {
      height: 15px;
    }
  `,

  large: css`
    font-size: 1.5rem;
    padding: 8px 13px;

    svg {
      height: 20px;
    }
  `,

  buttonType: css`
    background: ${({ theme, buttonType }) => theme.colors[buttonType]};
    color: ${({ theme, buttonType }) => theme.colors[withOnText(buttonType)]};

    &:hover {
      background: ${({ theme, buttonType }) => theme.colors.darker(1, buttonType)};
    }
  `,
};

const StyledButton = styled.button`
  border-radius: ${({ theme }) => theme.sizes.borderRadius};
  padding: ${({ theme }) => theme.sizes.spacingInput};
  background: ${({ theme }) => theme.colors.secondary};
  border: none;
  color: ${({ theme }) => theme.colors.onSecondary};
  cursor: pointer;
  font-size: 1.3rem;
  border: 2px solid
    ${({ theme, buttonType }) =>
      buttonType ? theme.colors[buttonType] : theme.colors.secondary};
  display: flex;
  align-items: center;
  display: inline-block;

  &:hover {
    background: ${({ theme }) => theme.colors.darker(1, 'secondary')};
    border: 2px solid
      ${({ theme, buttonType }) =>
        buttonType
          ? theme.colors.darker(1, buttonType)
          : theme.colors.darker(1, 'secondary')};
  }

  svg {
    height: 17px;
  }

  & > *:not(:last-child) {
    margin-right: 6px;
  }

  ${(props) => props.buttonType && propsCSS.buttonType};
  ${(props) => props.size === 'small' && propsCSS.small};
  ${(props) => props.size === 'large' && propsCSS.large};
  ${(props) => props.outline && propsCSS.outline};
`;

const Button = ({ children, ...buttonProps }) => (
  <StyledButton type="button" {...buttonProps}>
    {children}
  </StyledButton>
);

Button.propTypes = {
  children: PropTypes.node.isRequired,
  onClick: PropTypes.func.isRequired,
  buttonType: PropTypes.oneOf(['error', 'success', 'warning', 'info']),
  size: PropTypes.oneOf(['small', 'normal', 'large']),
  outline: PropTypes.bool,
};

Button.defaultProps = {
  type: 'info',
  size: 'normal',
  outline: false,
};

export default Button;
