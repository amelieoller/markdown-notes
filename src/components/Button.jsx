import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

const StyledButton = styled.button`
  background: white;
  color: ${props => props.theme.primaryHighlight};
  font-size: ${props => (props.small ? '.8em' : '1em')};
  padding: ${props => (props.small ? '0.1em .4em' : '0.25em 1em')};
  border: 2px solid ${props => props.theme.primaryHighlight};
  border-radius: 3px;
  cursor: pointer;

  &:hover,
  &:active {
    transition: all 0.3s ease;
    background-color: ${props => props.theme.primaryHighlight};
    color: white;
  }
`;

const Button = (props) => {
  const { text, onClick } = props;
  return (
    <StyledButton type="submit" onClick={() => onClick()} {...props}>
      {text}
    </StyledButton>
  );
};

Button.defaultProps = {
  text: 'Submit',
};

Button.propTypes = {
  text: PropTypes.string,
};

export default Button;
