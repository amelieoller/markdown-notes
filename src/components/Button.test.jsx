import React from 'react';
import { shallow, configure, mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import Button from './Button';

configure({ adapter: new Adapter() });

describe('<Button />', () => {
  it('renders', () => {
    const mockFn = jest.fn();

    shallow(<Button text="Button Text" onClick={mockFn} />);
  });

  it('matches the snapshot', () => {
    const mockFn = jest.fn();
    const wrapper = mount(<Button text="Button Text" onClick={mockFn} />);

    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('updates via props', () => {
    const mockFn = jest.fn();
    const wrapper = shallow(<Button text="Button Text" onClick={mockFn} />);

    expect(toJSON(wrapper)).toMatchSnapshot();
    wrapper.setProps({ text: 'New Button Text' });
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('should call mock function when button is clicked', () => {
    const mockFn = jest.fn();
    const wrapper = shallow(<Button text="Button Text" onClick={mockFn} />);

    wrapper.simulate('click');
    expect(mockFn).toHaveBeenCalled();
  });
});
