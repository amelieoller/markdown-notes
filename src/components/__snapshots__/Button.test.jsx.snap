// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> matches the snapshot 1`] = `
<Button
  onClick={[MockFunction]}
  text="Button Text"
>
  <styled.button
    onClick={[MockFunction]}
    text="Button Text"
    type="submit"
  >
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bdVaJa",
            "isStatic": false,
            "lastClassName": "dNdkMm",
            "rules": Array [
              "
  background: white;
  color: ",
              [Function],
              ";
  font-size: ",
              [Function],
              ";
  padding: ",
              [Function],
              ";
  border: 2px solid ",
              [Function],
              ";
  border-radius: 3px;
  cursor: pointer;

  &:hover,
  &:active {
    transition: all 0.3s ease;
    background-color: ",
              [Function],
              ";
    color: white;
  }
",
            ],
          },
          "displayName": "styled.button",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bdVaJa",
          "target": "button",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
      onClick={[MockFunction]}
      text="Button Text"
      type="submit"
    >
      <button
        className="sc-bdVaJa dNdkMm"
        onClick={[MockFunction]}
        type="submit"
      >
        Button Text
      </button>
    </StyledComponent>
  </styled.button>
</Button>
`;

exports[`<Button /> updates via props 1`] = `
<styled.button
  onClick={[MockFunction]}
  text="Button Text"
  type="submit"
>
  Button Text
</styled.button>
`;

exports[`<Button /> updates via props 2`] = `
<styled.button
  onClick={[MockFunction]}
  text="New Button Text"
  type="submit"
>
  New Button Text
</styled.button>
`;
